target/linux/mvebu/config-5.4

```diff
@@ -0,0 +1,526 @@
+CONFIG_HAVE_COPY_THREAD_TLS=y
+CONFIG_HAVE_C_RECORDMCOUNT=y
+CONFIG_HAVE_DEBUG_KMEMLEAK=y
+CONFIG_HAVE_DMA_CONTIGUOUS=y
+CONFIG_HAVE_DYNAMIC_FTRACE=y
+CONFIG_HAVE_DYNAMIC_FTRACE_WITH_REGS=y
+CONFIG_HAVE_EBPF_JIT=y
+CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS=y
+CONFIG_HAVE_FTRACE_MCOUNT_RECORD=y
+CONFIG_HAVE_FUNCTION_GRAPH_TRACER=y
+CONFIG_HAVE_FUNCTION_TRACER=y
+CONFIG_HAVE_IDE=y
+CONFIG_HAVE_IRQ_TIME_ACCOUNTING=y
+CONFIG_HAVE_LD_DEAD_CODE_DATA_ELIMINATION=y
+CONFIG_HAVE_MOD_ARCH_SPECIFIC=y
+CONFIG_HAVE_NET_DSA=y
+CONFIG_HAVE_OPROFILE=y
+CONFIG_HAVE_OPTPROBES=y
+CONFIG_HAVE_PCI=y
+CONFIG_HAVE_PERF_EVENTS=y
+CONFIG_HAVE_PERF_REGS=y
+CONFIG_HAVE_PERF_USER_STACK_DUMP=y
+CONFIG_HAVE_PROC_CPU=y
+CONFIG_HAVE_REGS_AND_STACK_ACCESS_API=y
+CONFIG_HAVE_RSEQ=y
+CONFIG_HAVE_SMP=y
+CONFIG_HAVE_SYSCALL_TRACEPOINTS=y
+CONFIG_HAVE_UID16=y
+CONFIG_HAVE_VIRT_CPU_ACCOUNTING_GEN=y
+CONFIG_HIGHMEM=y
+# CONFIG_HIGHPTE is not set
+CONFIG_HOTPLUG_CPU=y
+CONFIG_HWBM=y
+CONFIG_HWMON=y
+CONFIG_HW_RANDOM=y
+CONFIG_HZ_FIXED=0
+CONFIG_HZ_PERIODIC=y
+CONFIG_I2C=y
+CONFIG_I2C_BOARDINFO=y
+CONFIG_I2C_CHARDEV=y
+CONFIG_I2C_MV64XXX=y
+# CONFIG_I2C_PXA is not set
+CONFIG_INITRAMFS_SOURCE=""
+CONFIG_IRQCHIP=y
+CONFIG_IRQ_DOMAIN=y
+CONFIG_IRQ_DOMAIN_HIERARCHY=y
+CONFIG_IRQ_FORCED_THREADING=y
+CONFIG_IRQ_WORK=y
+# CONFIG_IWMMXT is not set
+CONFIG_JBD2=y
+# CONFIG_KASAN_STACK is not set
+CONFIG_LEDS_GPIO=y
+CONFIG_LEDS_PCA963X=y
+CONFIG_LEDS_TLC591XX=y
+CONFIG_LEDS_TRIGGER_DISK=y
+CONFIG_LIBFDT=y
+CONFIG_LOCK_DEBUGGING_SUPPORT=y
+CONFIG_LOCK_SPIN_ON_OWNER=y
+CONFIG_LZO_COMPRESS=y
+CONFIG_LZO_DECOMPRESS=y
+CONFIG_MACH_ARMADA_370=y
+# CONFIG_MACH_ARMADA_375 is not set
+CONFIG_MACH_ARMADA_38X=y
+# CONFIG_MACH_ARMADA_39X is not set
+CONFIG_MACH_ARMADA_XP=y
+# CONFIG_MACH_DOVE is not set
+CONFIG_MACH_MVEBU_ANY=y
+CONFIG_MACH_MVEBU_V7=y
+CONFIG_MAGIC_SYSRQ=y
+CONFIG_MANGLE_BOOTARGS=y
+CONFIG_MARVELL_PHY=y
+CONFIG_MDIO_BUS=y
+CONFIG_MDIO_DEVICE=y
+CONFIG_MDIO_I2C=y
+CONFIG_MEMFD_CREATE=y
+CONFIG_MEMORY=y
+CONFIG_MIGHT_HAVE_CACHE_L2X0=y
+CONFIG_MIGRATION=y
+CONFIG_MMC=y
+CONFIG_MMC_BLOCK=y
+CONFIG_MMC_MVSDIO=y
+CONFIG_MMC_SDHCI=y
+# CONFIG_MMC_SDHCI_PCI is not set
+CONFIG_MMC_SDHCI_PLTFM=y
+CONFIG_MMC_SDHCI_PXAV3=y
+# CONFIG_MMC_TIFM_SD is not set
+CONFIG_MODULES_USE_ELF_REL=y
+CONFIG_MTD_CFI_STAA=y
+CONFIG_MTD_NAND_CORE=y
+CONFIG_MTD_NAND_ECC_SW_HAMMING=y
+CONFIG_MTD_NAND_MARVELL=y
+CONFIG_MTD_RAW_NAND=y
+CONFIG_MTD_SPI_NOR=y
+CONFIG_MTD_SPLIT_FIRMWARE=y
+CONFIG_MTD_UBI=y
+CONFIG_MTD_UBI_BEB_LIMIT=20
+CONFIG_MTD_UBI_BLOCK=y
+# CONFIG_MTD_UBI_FASTMAP is not set
+# CONFIG_MTD_UBI_GLUEBI is not set
+CONFIG_MTD_UBI_WL_THRESHOLD=4096
+CONFIG_MUTEX_SPIN_ON_OWNER=y
+CONFIG_MVEBU_CLK_COMMON=y
+CONFIG_MVEBU_CLK_COREDIV=y
+CONFIG_MVEBU_CLK_CPU=y
+CONFIG_MVEBU_DEVBUS=y
+CONFIG_MVEBU_MBUS=y
+CONFIG_MVMDIO=y
+CONFIG_MVNETA=y
+CONFIG_MVNETA_BM=y
+CONFIG_MVNETA_BM_ENABLE=y
+# CONFIG_MVPP2 is not set
+CONFIG_MVSW61XX_PHY=y
+CONFIG_MV_XOR=y
+CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_NEON=y
```

Comment:
Why is this enabled?

Reply:
NEON instruction is supported by Armada 375+
If the kernel size is a concern (Linksys Mamba's kernel size is reaching its limit) , I will move Armada 380+ to a different subtarget, and disable NEON on this subtarget.

Reply:
With your last push, and going back to my original build for the mamba (gcc9, binutils2.34, O2) there appears to be ~20k remaining. Might make sense if NEON removal buys anything of consequence on this subtarget.

Reply:
> NEON instruction is supported by Armada 375+
> If the kernel size is a concern (Linksys Mamba's kernel size is reaching its limit) , I will move Armada 380+ to a different subtarget, and disable NEON on this subtarget.

The additional subtarget for NEON capable devices or activating it in gcc was rejected already, personally, don't think there's much to gain.
I don't know if kernel uses NEON extensions if compiler and userspace doesn't have it enabled, looking at lib-xor it probably does. So this symbol could stay enabled.

Reply:
AFAIK SIMD instructions such as NEON will improve cryptographic performance

@cotequeiroz Am I right?

Reply:
Using neon in kernel is not straightforward.  It comes with a few restrictions related to register file preserve/restore limitations.  Code must contained between `kernel_neon_begin` and `kernel_neon_end` calls, which are enabled by `CONFIG_NEON=y`.  So by itself it does not make general use of neon.  See [the kernel documentation](https://www.kernel.org/doc/Documentation/arm/kernel_mode_neon.rst) for more information.

Neon helps crypto a lot.  I haven't tried yet to run without it on my WRT3200ACM to point out speed gains, but on ipq40xx, use of Neon-optimized ghash module alone increases AES-GCM throughput by 40%, and it beats the ipq40xx crypto engine in network-sized AES128-GCM blocks (where the crypto engine does AES-CTR).  I've posted results for [AES-GCM](https://github.com/openwrt/openwrt/pull/2518#issuecomment-583532920), and for [other AES modes](https://github.com/openwrt/openwrt/pull/2518#issuecomment-590808885).  Use the 1024-bytes block size to gauge network performance.

Reply:
@cotequeiroz as I understand, kernel doesn't need compiler or userspace to have neon enabled. For example enabling kernel symbols would benefit WireGuard and whatever that uses kernel crypto but not OpenVPN or OpenSSL, and this would be less invasive, instead of enabling neon in gcc. Please do clarify if I'm understanding it wrong.

@dengqf6 please remove the addition of subtarget to another PR, this is clobbering discussion as @mans0n said. It averts the focus from main point of this PR, which could postpone merging for a long time.

Reply:
@tmn505 Done

Reply:
@tmn505 
Neon needs to be enabled by the kernel to be used.  If you have `CONFIG_NEON` unset, then the kernel will not enable it and any instruction use will generate an exception.  The symbol was added in [b5872db4a2](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=b5872db4a2ebe7dbc7a5e4013ae8ee37f3de3b97).  Setting CP10 and CP11 bits enables neon use from userspace.  This is done by setting `CONFIG_NEON`.  Without this, even if you use asm instructions, you won't be able to run them.
The next step is `CONFIG_KERNEL_MODE_NEON`.  It enables use of neon instructions from within the kernel.  This will enable acceleration through IPSEC, wireguard, and whatever access them from the kernel.  OpenSSL will not need `KERNEL_MODE_NEON` or gcc support to run them, since it uses asm directly (on by default), but it will need `CONFIG_NEON`.  That's why it can't be used to gauge any speedup from gcc neon use (unless you build it with `OPENSSL_WITH_ASM` turned off).  It was, IIRC, one of the pitfalls of deciding whether to enable gcc neon support or not in the past.
If you enable it in gcc, then it _**may**_ issue neon instructions.  Looking at the [Arm Documentation](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dht0002a/ch01s04s03.html), it does not appear to be straightforward for the compiler to identify parallelizing opportunities, so we may not gain much just by enabling in gcc without changing code.  Also, the floating port instructions are not used by default because it lacks handling of [denormal numbers](https://en.wikipedia.org/wiki/Denormal_number), so you'll get loss of precision.  You can override this with `-funsafe-math-optimizations`.  Who can guarantee that a program will work right 100% of the time?  Lossy compression algorithms should be fine, for example, but it has to be an exception, and even if the main purpose of a package is to provide lossy compression, there is still a chance that code will be affected it in other areas.  Granted, if we just keep neon disabled in gcc, then we'll get no benefit for sure.  My conclusion is, don't expect much gain in most cases.  If I can find some time, I'll build openssl without asm, and perhaps provide some numbers.

Reply:
Thanks, that clears things. Would be nice if someone with Armada 370/XP could test if there's no problems with CONFIG_NEON enabled.

Reply:
There is no change here from 4.19, which sets _CONFIG_KERNEL_MODE_NEON_ in target _generic_ and _CONFIG_NEON_ in target _mvebu_.